torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -431,84 +431,189 @@ c10::optional<at::Tensor> ComputeConstantFolding(Node* n, int opset_version) {
+  int minus_one_pos = -1;
+  for (int i = 0; i < shape.size(); ++i) {
+    if (shape[i] == -1) {
+      minus_one_pos = i;
+      break;
+    }
+  }
 
-  if (!reshape_has_zero && minus_one_pos == -1) {
-    return reshape;
+  if (!shape_has_zero && minus_one_pos == -1) {
+    return shape;
   }
   std::vector<int64_t> final_shape;
   // shape_ratio is used to calculate the real value of -1.
-  // One example with reshape 0 and -1:
+  // One example with shape 0 and -1:
   // input_shape = 2 16 5 4
-  // reshape = -1 0 4
+  // shape = -1 0 4
   // final_shape = 10 16 4
-  double shape_ratio = 1.0;
-  for (const auto i : c10::irange(input_shape_size)) {
-    shape_ratio *= static_cast<double>(input_shape[i]);
-  }
-  for (const auto i : c10::irange(reshape_size)) {
-    if (i != minus_one_pos) {
-      if (reshape[i] != 0) {
-        shape_ratio /= static_cast<double>(reshape[i]);
+  uint64_t shape_ratio = 1;
+  for (const auto& cur_input_shape : input_shape) {
+    shape_ratio *= static_cast<uint64_t>(cur_input_shape);
+  }
+  int shape_size = static_cast<int>(shape.size());
+  for (const auto i : c10::irange(shape_size)) {
+    if (i == minus_one_pos) {
+      continue;
+    }
+    if (shape[i] != 0) {
+      shape_ratio /= static_cast<uint64_t>(shape[i]);
+    } else {
+      shape_ratio /= static_cast<uint64_t>(input_shape[i]);
+    }
+  }
+
+  for (const int i : c10::irange(minus_one_pos)) {
+    int64_t cur_shape = shape[i] == 0 ? input_shape[i] : shape[i];
+    final_shape.push_back(cur_shape);
+  }
+  if (minus_one_pos != -1) {
+    final_shape.push_back(static_cast<int64_t>(std::round(shape_ratio)));
+  }
+  for (auto i = minus_one_pos + 1; i < shape_size; i++) {
+    int64_t cur_shape = shape[i] == 0 ? input_shape[i] : shape[i];
+    final_shape.push_back(cur_shape);
+  }
+  return final_shape;
+}
+
+// Similar to the function above, but for symbolic shapes.
+c10::optional<::c10::SymbolicShape> ComputeShapeFromReshape(
+    Node* n,
+    const c10::SymbolicShape& input_shape,
+    const c10::SymbolicShape& shape,
+    int opset_version) {
+  std::vector<c10::ShapeSymbol> input_shape_vector =
+      input_shape.sizes().value();
+  std::vector<c10::ShapeSymbol> shape_vector = shape.sizes().value();
+  TORCH_INTERNAL_ASSERT(
+      !input_shape_vector.empty() || !shape_vector.empty(),
+      "Reshape node should have at least one input size > 0 when constant folding.");
+  if (shape_vector.empty()) {
+    return input_shape;
+  }
+  if (input_shape_vector.empty()) {
+    return shape;
+  }
+
+  auto is_zero = [](c10::ShapeSymbol& ss) { return ss.value() == 0; };
+  auto it_0 = std::find_if(shape_vector.begin(), shape_vector.end(), is_zero);
+  bool shape_has_zero = it_0 != shape_vector.end();
+
+  if (opset_version >= 14 && n->hasAttributeS("allowzero")) {
+    int allowzero = n->i(attr::allowzero);
+    if (allowzero == 1 && shape_has_zero) {
+      // TODO: Shall we handle the case when shape has -1 here?
```

Comment:
Please expand the comment explaining the divide by zero issue.

But would it be correct to just not multiply the shape_ratio if a particular axis is 0?
Something like:
`if (input_shape.is_static() && input_shape.static_size()) {`

On line 564?

Reply:
The issue here is divide by zero, so this is the case there are 0 in denominator (shape), not the input, so skipping multiply cannot resolve the issue.

Reply:
Please add a bit more context to the comment here so the TODO is clear to someone else in the distant future.

Reply:
Seems that you are looking at the old code, the new comments added more info there:

      // Return here, because the denominator in shape_ratio calculation cannot
      // be 0.
      // TODO: Shall we handle the case when shape has -1 here?